OpenRA.Mods.Common/Traits/DeliversExperience.cs

```diff
@@ -0,0 +1,127 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Give experience actor has to other actors with the `AcceptsDeliveredExperience` trait.")]
+	class DeliversExperienceInfo : ITraitInfo, Requires<GainsExperienceInfo>
+	{
+		[Desc("The amount of experience the donating player receives.")]
+		public readonly int PlayerExperience = 0;
+
+		[Desc("Identifier checked against AcceptsDeliveredExperience.ValidTypes. Only needed if the latter is not empty.")]
+		public readonly string Type = null;
+
+		[VoiceReference] public readonly string Voice = "Action";
+
+		public object Create(ActorInitializer init) { return new DeliversExperience(this); }
+	}
+
+	class DeliversExperience : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly DeliversExperienceInfo info;
+
+		public DeliversExperience(DeliversExperienceInfo info)
+		{
+			this.info = info;
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get { yield return new DeliversExperienceOrderTargeter(); }
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (target.Actor.Trait<GainsExperience>().Level == target.Actor.Trait<GainsExperience>().MaxLevel || self.Trait<GainsExperience>().Level == 0)
```

Comment:
@MustaphaTR You can pass self through the constructor when you call it in
`public object Create(ActorInitializer init) { return new DeliversExperience(this); }`
and then cache it, I am not saying what @GraionDilach is better since I didn't look but this is how you can do it.